sbncode/SinglePhotonAnalysis/HelperFunctions/helper_math.h

```diff
@@ -0,0 +1,295 @@
+                        
+                    //std::cout<<"KNK: "<<bc<<" "<<hs[0]<<" "<<result.start().X()<<" "<<result.start().Y()<<" "<<result.start().Z()<<" "<<result.end().X()<<" "<<result.end().Y()<<" "<<result.end().Z()<<std::endl; 
+//                    std::cout<<wireids[0].Plane<<" "<<result.start().X()<<std::endl;
+                    if(plane != (int)wireids[0].Plane) continue;
+
+        // for the dead wires on the same plane 
+                    std::vector<double> start = {0.0,result.start().Y(),result.start().Z()};
+                    std::vector<double> end = {0.0,result.end().Y(),result.end().Z()};
+                    std::vector<double> point = {0.0,Ypoint,Zpoint};
+                    double dist = dist_line_point(start,end,point);
+                    min_dist = std::min(dist,min_dist);
+             }
+                 
+    return min_dist;
+
+    }
+
+//--------------- end of copying from bad_channel_matching.h
+
+    double impact_paramater_shr(double x, double y, double z, art::Ptr<recob::Shower> & shr){
+
+        std::vector<double> vert = {x,y,z}; 
+        std::vector<double> start = {shr->ShowerStart().X(), shr->ShowerStart().Y(),shr->ShowerStart().Z()};
+        std::vector<double> abit = {shr->ShowerStart().X() + shr->Direction().X(),  shr->ShowerStart().Y()+shr->Direction().Y(),  shr->ShowerStart().Z()+shr->Direction().Z()};
+
+        return dist_line_point(start, abit, vert);
+
+    }
+
+    // invariant mass of a particle that decays to two showers
+    double  implied_invar_mass(double vx, double vy, double vz, art::Ptr<recob::Shower> & s1, double E1,  art::Ptr<recob::Shower> &s2, double E2){
+
+        double s1x = s1->ShowerStart().X()-vx;
+        double s1y = s1->ShowerStart().Y()-vy;
+        double s1z = s1->ShowerStart().Z()-vz;
+        double norm1  = sqrt(pow(s1x,2)+pow(s1y,2)+pow(s1z,2));
+        s1x = s1x/norm1; //unit vector pointing to shower start from point (vx, vy,vz)
+        s1y = s1y/norm1;
+        s1z = s1z/norm1;
+
+        double s2x = s2->ShowerStart().X()-vx;
+        double s2y = s2->ShowerStart().Y()-vy;
+        double s2z = s2->ShowerStart().Z()-vz;
+        double norm2  = sqrt(pow(s2x,2)+pow(s2y,2)+pow(s2z,2));
+        s2x = s2x/norm2; // unit vector pointing to shower start from point (vx, vy, vz)
+        s2y = s2y/norm2;
+        s2z = s2z/norm2;
+
+        return sqrt(2.0*E1*E2*(1.0-(s1x*s2x+s1y*s2y+s1z*s2z)));
+
+
+    }
+
+    // invariant mass of two showers, calculated directly from shower directions
+    double  invar_mass(art::Ptr<recob::Shower> & s1, double E1,  art::Ptr<recob::Shower> &s2, double E2){
+
+        double s1x = s1->Direction().X();
+        double s1y = s1->Direction().Y();
+        double s1z = s1->Direction().Z();
+
+        double s2x = s2->Direction().X();
+        double s2y = s2->Direction().Y();
+        double s2z = s2->Direction().Z();
+
+        return sqrt(2.0*E1*E2*(1.0-(s1x*s2x+s1y*s2y+s1z*s2z)));
```

Comment:
Do the vector objects not provide a dot product operation?

Reply:
The vector calculations implemented in `helper_math.*` & `SEAview/SEAviewer.h` can further be simplified using functions from `std` or ROOT:
 `std::inner_product(a.begin(), a.end(), b.begin(), 0);` or `<TVector> a; <TVector> b; a.Dot(b)`;

Perhaps we can further clean them up by merging or factoring similar calculations. Since the functionality is still the same before/after applying the change, I may save it for next update.